### ISSUE TITLE
es3 safe mode

### ISSUE THREAD

**Issue by stefanpenner** [esperantojs/esperanto#84]
While upgrading ember-cli users to esperanto, we realized that the correct cycle protection offered by defineProperty unfortunately is un-available for our poor ES3 serving comrades. 

As it turns out, in many cases the defProp isn't actually required, but in the pathalogical cycle case its lazyiness is works in our favor.  Although suboptimal, allowing an opt-out, allows our ES3 users to continue the same, or better semantics then before.

https://github.com/esperantojs/esperanto/blob/1cd4e4c7bf81d9a4634c595e0f7f44fa6a954535/src/standalone/builders/strictMode/utils/transformBody.js#L86

Thoughts?

In theory, adding a validation phase, could still assure program correctness, throwing if a non es3 safe appears.

**Comment by Rich-Harris** [comment]
> in many cases

That phrase rings alarm bells to me! I think having an ES3 mode that occasionally resulted in different behaviour would cause subtle bugs to appear at the worst moments.

Re-exporting another module's bindings seems to me like a niche use case (though everything seems like a niche use case if you haven't done it yourself...), so I'm inclined to suggest that an ES3 'safe mode' should simply throw an error if an ES6 module is doing something that can't be represented in ES3. (Of course, it would be a helpful error, that e.g. pointed to a wiki page explaining the problem.) It feels like the support burden is then being placed where it belongs - on the people choosing to use (most of the) ES6 module features.

So in summary I prefer the validation suggestion. How does that sound?

**Comment by eventualbuddha** [comment]
> I'm inclined to suggest that an ES3 'safe mode' should simply throw an error if an ES6 module is doing something that can't be represented in ES3

That sounds right to me.

**Comment by stefanpenner** [comment]
> so I'm inclined to suggest that an ES3 'safe mode' should simply throw an error if an ES6 module is doing something that can't be represented in ES3.

this is good, it does require analysis of the entirely module graph. Many re-export examples, are inherently safe without laziness.

> That phrase rings alarm bells to me! I think having an ES3 mode that occasionally resulted in different behaviour would cause subtle bugs to appear at the worst moments.

The reality is, consumers of the older module transpiler, have found a fairly broad success path that is now entirely gone for ES3  environments.

**Comment by Rich-Harris** [comment]
As things stand, the one-to-one transpiler knows nothing about imported modules, so it can't know whether it's safe to do `exports.foo = imported.foo` as opposed to the `defineProperty` stuff. In the context of a bundle, we presumably *could* make that determination, and modify [this block](https://github.com/esperantojs/esperanto/blob/master/src/bundler/combine/transformBody.js#L83-L97) accordingly.

I hear your point though, it is a shame if people targeting ES3 were already re-exporting bindings and now can't. How widespread is that? Am I right in thinking that (and namespace exports within a bundle) are the only casualties of ES3? If it's only a small problem that's solved by simply importing from the original exporter, how much of a headache would it cause to fix that on an app-by-app basis? (Not a rhetorical question.)

> Ultimately this whole problem makes me sad, hopefully IE8 and friends leave us soon and we can live much nicer lives.

Amen.

**Comment by stefanpenner** [comment]
> I hear your point though, it is a shame if people targeting ES3 were already re-exporting bindings and now can't. How widespread is that? Am I right in thinking that (and namespace exports within a bundle) are the only casualties of ES3? If it's only a small problem that's solved by simply importing from the original exporter, how much of a headache would it cause to fix that on an app-by-app basis? (Not a rhetorical question.)

I would still like to add a `__very_evil_es3_safe: true` flag, and handle the validation phase externally

**Comment by stefanpenner** [comment]
> How widespread is that?

unfortunately :(

**Comment by Rich-Harris** [comment]
ember.js? what's that, is it a project I should know about? :P

**Comment by stefanpenner** [comment]
@Rich-Harris can I add the flag?
i really want to prevent us from doing https://github.com/rwjblue/emberjs-build/blob/esperanto/lib/utils/transpile-es6.js#L5-L17 for our IE8 users..

**Comment by Rich-Harris** [comment]
Yeah, I don't see why not. As long as it's a caveat emptor thing, then I guess we've done our duty.

**Comment by stefanpenner** [comment]
i was thinking something like `___I_KNOW_THIS_IS_BAD: true`

**Comment by Rich-Harris** [comment]
works for me :)

**Comment by stefanpenner** [comment]
:+1:
